package/MDAnalysis/analysis/backends.py

```diff
@@ -0,0 +1,182 @@
+:class:`BackendMultiprocessing` that supports parallelization via standard
+python ``multiprocessing`` module, and :class:`BackendDask`, that uses the same
+process-based parallelization as :class:`BackendMultiprocessing`, but different
+serialization algorithm.
+==============================================================
+"""
+import warnings
+from typing import Callable
+from MDAnalysis.lib.util import is_installed
+
+
+class BackendBase:
+    """Base class for backend implementation. Initializes an instance and performs
+    checks for its validity, such as n_workers and possibly other ones.
+
+    Parameters
+    ----------
+    n_workers : int
+        number of workers (usually, processes) over which the work is split
+
+    Examples
+    --------
+    .. code-block:: python
+        # implement a thread-based backend
+        from MDAnalysis.analysis.backends import BackendBase
+        class ThreadsBackend(BackendBase):
+            def apply(self, func, computations):
+                from multiprocessing.dummy import Pool
+
+                with Pool(processes=self.n_workers) as pool:
+                    results = pool.map(func, computations)
+                return results
+        from MDAnalysis.analysis.rms import RMSD
+        R = RMSD(...) # setup the run
+        n_workers = 2
+        backend = ThreadsBackend(n_workers=n_workers)
+        R.run(backend=backend)
+
+    .. versionadded:: 2.8.0
+    """
+
+    def __init__(self, n_workers: int):
+        self.n_workers = n_workers
+        self._validate()
+
+    def _get_checks(self):
+        """Get dictionary with `condition: error_message` pairs that ensure the
+        validity of the backend instance
+
+        Returns
+        -------
+        dict
+            dictionary with `condition: error_message` pairs that will get
+            checked during _validate() run
+
+        .. versionadded:: 2.8.0
+        """
+        return {
+            isinstance(self.n_workers, int) and self.n_workers > 0:
+            f"n_workers should be positive integer, got {self.n_workers=}",
+        }
+
+    def _get_warnings(self):
+        """Get dictionary with `condition: warning_message` pairs that ensure
+        the good usage of the backend instance
+
+        Returns
+        -------
+        dict
+            dictionary with `condition: warning_message` pairs that will get
+            checked during _validate() run
+
+        .. versionadded:: 2.8.0
+        """
+        return dict()
+
+    def _validate(self):
+        """Check correctness (e.g. `dask` is installed if using `backend='dask'`)
+        and good usage (e.g. `n_workers=1` if backend is serial) of the backend
+
+        Raises
+        ------
+        ValueError
+            if one of the conditions in :meth:`self._get_checks()` is True
+
+        .. versionadded:: 2.8.0
+        """
+        for check, msg in self._get_checks().items():
+            if not check:
+                raise ValueError(msg)
```

Comment:
```suggestion
                raise RuntimeError(msg)
```
Not sure if `ValueError` is very informative in all cases here, maybe `RuntimeError` is a better exception for this purpose?

Reply:
In defense of @marinegor  I am probably someone who said during earlier reviews that [ValueErrors](https://docs.python.org/3/library/exceptions.html#ValueError) should be raised when a value is not correct, e.g., because it's not available as a backend. If you choose [RuntimeError](https://docs.python.org/3/library/exceptions.html#RuntimeError) (which is "raised when an error is detected that doesn’t fall in any of the other categories") the msg should state clearly what went wrong.

Reply:
That's very fair. My line of thought here is that the message is indeed very informative. If the checks end up checking stuff other than parameters, ValueError might not be general enough. Just thinking out loud, since exceptions are technically user-facing.

Reply:
I would stick to `ValueError`, but thanks!

Reply:
> If the checks end up checking stuff other than parameters, `ValueError` might not be general enough.

I agreed with @orbeckst and @marinegor until you pointed this out @RMeli.
Perhaps the `_get_checks` and `_get_warnings` functions could just be a series of checks that raise an appropriate exception / warning when necessary:

```python
def _get_checks(self):
    """Check correctness (e.g. ``dask`` is installed if using ``backend='dask'``)
     of the backend
    Raises
    ------
    ValueError
        if `n_workers` is not a positive integer
    """
    if not isinstance(self.n_workers, int) or self.n_workers <= 0:
        raise ValueError(f"n_workers should be positive integer, got {self.n_workers=}")
```

And `_validate` would then look like this:

```python
def _validate(self):
    """Check correctness and good usage of the backend"""
    self._get_checks()
    self._get_warnings()
```

This way, the most appropriate exception can be raised, and it's more explicit what is being validated

Reply:
@p-j-smith I think that was my first idea, but then I thought that it could be good to be able to get the list of warning/exceptions without necessarily raising them -- for example, when trying to configure some more complex backend, like on a cluster.

But it could also be that I'm solving a problem that doesn't yet exist.

Reply:
Oh I see, so someone could use this function to know how they should configure their backend? I think the usual way to do this would be with docs and tutorials showing to configure the backend

Reply:
>Oh I see, so someone could use this function to know how they should configure their backend?

Yes, more or less -- or try to configure it few times without explicitly catching an exception.

>I think the usual way to do this would be with docs and tutorials showing to configure the backend

I don't mind changing it, but still think that this way is slightly more flexible to the potential developer.

Reply:
Can we consider this resolved, i.e. the code will stay as-is?

Reply:
I'm ok with that

Reply:
I'm ok with that

Reply:
LGTM